### PR TITLE
Abi encoder decoder

### PR DESCRIPTION
**Pull request by billettc**

### CODE REVIEW

**Comment by maoueh** [comment]
Remains: Test a Smart Contract that output in string the following types:

* asset
* extended_asset
* symbol
* symbol_code

Simple way is to create a contract that saves these four values in a table and then, print the table.

File: abiencoder.go [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+			if e := binaryEncoder.writeByte(1); e != nil {
+				return e
+			}
+		} else {
+			Logger.ABIEncoder.Printf("Field [%s] of type [%s] is option and not present\n", fieldName, fieldType)
+			return binaryEncoder.writeByte(0)
+		}
+
+	} else if !value.Exists() {
+		return fmt.Errorf("encode field: none optional field [%s] as a nil value", fieldName)
```
[the remark this answers is not in the document]
Comment (billettc) [reply abiencoder.go]: No need to log when and error occurs. Let the caller handle the err and log it.

File: abiencoder.go [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+	case "block_timestamp_type":
+		time, err := time.Parse("2006-01-02T15:04:05.999999-07:00", value.Str)
+		if err != nil {
+			return fmt.Errorf("writing field: block_timestamp_type: %s", err)
+		}
+		object = BlockTimestamp{
+			Time: time,
+		}
+	case "name":
+		if len(value.Str) > 12 {
```
[the remark this answers is not in the document]
Comment (billettc) [reply abiencoder.go]: Nope,

**Comment by maoueh** [comment]
@billettc: 

```
    {
      "id": 1,
      "asset": "0.0000 EOS",
      "extended_asset": {
        "quantity": "0.0000 EOS",
        "contract": "register1"
      },
      "symbol": "4,EOS",
      "symbol_code": "EOS",
      "block_timestamp": "2000-01-01T00:00:00.000",
      "time_point": "1970-01-01T00:00:00.000",
      "time_point_sec": "1970-01-01T00:00:00",
      "signature": "",
      "public_key": "EOS4w8ifWypqDq7rtg2SkhPvJ1h89NuDfhs3VCir5UTz3JZmvZ6D8",
      "checksum256": "0000000000000000000000000000000000000000000000000000000000000000",
      "bytes": ""
    }
```

I was unable to make `bytes` and "signature" to generate something. Also, my code is not generating correctly a `public_key`. But I double checked with `eosc get table eosio eosio producers` and it's the correct output, so I included it as a sample in the snippet above.

Otherwise, for future reference:

```
# In .hpp, struct row + table definition
#include <eosiolib/symbol.hpp>
#include <eosiolib/eosio.hpp>
#include <eosiolib/asset.hpp>
#include <eosiolib/time.hpp>
#include <eosiolib/transaction.hpp>

...

 struct test {
            uint64_t              id;
            eosio::asset                 asset;
            eosio::extended_asset        extended_asset;
            eosio::symbol_type           symbol;
            uint64_t                     symbol_code;
            eosio::block_timestamp       block_timestamp;
            eosio::time_point            time_point;
            eosio::time_point_sec        time_point_sec;
            signature                    signature;
            public_key                   public_key;
            checksum256                  checksum256;
            eosio::bytes                 bytes;

            auto primary_key()const { return id; }
        };

        typedef eosio::multi_index<
            N(test), test
        > tests;
```

```
# In .cpp for row emplace
    tests test_table(_self, _self);
    test_table.emplace(account, [&](test& row) {
        row.id = test_table.available_primary_key();
        row.asset = eosio::asset();
        row.extended_asset = eosio::extended_asset(eosio::asset(), account);
        row.symbol = eosio::asset().symbol;
        row.symbol_code = eosio::asset().symbol.name();
        row.block_timestamp = eosio::block_timestamp();
        row.time_point = eosio::time_point();
        row.time_point_sec = eosio::time_point_sec();
        row.signature = signature();
        row.public_key = public_key();
        row.checksum256 = checksum256();
        row.bytes.push_back(0x80);
        row.bytes.push_back(0x81);
    });
```

```
# In ABI file for row
    {
      "name": "test_row",
      "base": "",
      "fields": [
        { "name": "id", "type": "uint64" },
        { "name": "asset", "type": "asset" },
        { "name": "extended_asset", "type": "extended_asset" },
        { "name": "symbol", "type": "symbol" },
        { "name": "symbol_code", "type": "symbol_code" },
        { "name": "block_timestamp", "type": "block_timestamp_type" },
        { "name": "time_point", "type": "time_point" },
        { "name": "time_point_sec", "type": "time_point_sec" },
        { "name": "signature", "type": "bytes" },
        { "name": "public_key", "type": "bytes" },
        { "name": "checksum256", "type": "checksum256" },
        { "name": "bytes", "type": "bytes" }
      ]
    }
```

**Comment by abourget** [comment]
Doesn't support arrays of structs yet. Ex:
```
{
  "account_name": "eospixelsd14",
  "abi": {
    "version": "eosio::abi/1.0",
    "types": [
      {
        "new_type_name": "time",
        "type": "uint32"
      }
    ],
    "structs": [
      {
        "name": "canvas",
        "base": "",
        "fields": [
          {
            "name": "id",
            "type": "uint64"
          },
          {
            "name": "lastPaintedAt",
            "type": "uint64"
          },
          {
            "name": "duration",
            "type": "uint32"
          }
        ]
      },
      {
        "name": "pixel",
        "base": "",
        "fields": [
          {
            "name": "color",
            "type": "uint32"
          },
          {
            "name": "priceCounter",
            "type": "uint16"
          },
          {
            "name": "owner",
            "type": "name"
          }
        ]
      },
      {
        "name": "pixel_row",
        "base": "",
        "fields": [
          {
            "name": "row",
            "type": "uint64"
          },
          {
            "name": "pixels",
            "type": "pixel[]"
          }
        ]
      },
      {
        "name": "account",
        "base": "",
        "fields": [
          {
            "name": "owner",
            "type": "name"
          },
          {
            "name": "balance",
            "type": "asset"
          }
        ]
      },
      {
        "name": "init",
        "base": ""
      },
      {
        "name": "refresh",
        "base": ""
      },
      {
        "name": "changedur",
        "base": "",
        "fields": [
          {
            "name": "duration",
            "type": "time"
          }
        ]
      },
      {
        "name": "end",
        "base": ""
      },
      {
        "name": "createacct",
        "base": "",
        "fields": [
          {
            "name": "account",
            "type": "name"
          }
        ]
      },
      {
        "name": "withdraw",
        "base": "",
        "fields": [
          {
            "name": "to",
            "type": "name"
          },
          {
            "name": "quantity",
            "type": "asset"
          }
        ]
      },
      {
        "name": "clearpixels",
        "base": "",
        "fields": [
          {
            "name": "count",
            "type": "uint16"
          },
          {
            "name": "nonce",
            "type": "uint16"
          }
        ]
      },
      {
        "name": "clearaccts",
        "base": "",
        "fields": [
          {
            "name": "count",
            "type": "uint16"
          },
          {
            "name": "nonce",
            "type": "uint16"
          }
        ]
      },
      {
        "name": "clearcanvs",
        "base": "",
        "fields": [
          {
            "name": "count",
            "type": "uint16"
          },
          {
            "name": "nonce",
            "type": "uint16"
          }
        ]
      }
    ],
    "actions": [
      {
        "name": "init",
        "type": "init",
        "ricardian_contract": ""
      },
      {
        "name": "refresh",
        "type": "refresh",
        "ricardian_contract": ""
      },
      {
        "name": "changedur",
        "type": "changedur",
        "ricardian_contract": ""
      },
      {
        "name": "end",
        "type": "end",
        "ricardian_contract": ""
      },
      {
        "name": "createacct",
        "type": "createacct",
        "ricardian_contract": ""
      },
      {
        "name": "withdraw",
        "type": "withdraw",
        "ricardian_contract": ""
      },
      {
        "name": "clearpixels",
        "type": "clearpixels",
        "ricardian_contract": ""
      },
      {
        "name": "clearaccts",
        "type": "clearaccts",
        "ricardian_contract": ""
      },
      {
        "name": "clearcanvs",
        "type": "clearcanvs",
        "ricardian_contract": ""
      }
    ],
    "tables": [
      {
        "name": "canvases",
        "index_type": "i64",
        "key_names": [
          "id"
        ],
        "key_types": [
          "uint64"
        ],
        "type": "canvas"
      },
      {
        "name": "pixels",
        "index_type": "i64",
        "key_names": [
          "row"
        ],
        "key_types": [
          "uint64"
        ],
        "type": "pixel_row"
      },
      {
        "name": "account",
        "index_type": "i64",
        "key_names": [
          "owner"
        ],
        "key_types": [
          "name"
        ],
        "type": "account"
      }
    ]
  }
}
```

See the `pixels` in `pixel_row` .. of type `pixel[]`.

Errors out with: ` decoding fields: decoding field [pixels] of type [pixel[]]: read field of type [pixel[]]: unknown type`

**Pull request merged**